### PR TITLE
Added Pagerfanta Trait

### PR DESCRIPTION
**Pull request by MarioBlazek**
This PR adds shorthand method for creating Pagerfanta pagers.

Should reconsider:
* [x] method and trait naming
* [x] tests
* [x] updating methods documentation

### CODE REVIEW

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+     *
+     * @param AdapterInterface $adapter
+     * @param Request $request
+     * @param int $maxPerPage
+     *
+     * @return Pagerfanta
+     */
+    protected function getPagerfantaPager(AdapterInterface $adapter, Request $request, $maxPerPage)
+    {
+        $pager = new Pagerfanta($adapter);
```
Comment (emodric) [inline lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: This should also have `$pager->setNormalizeOutOfRangePages(true);`

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        $pager->setMaxPerPage($maxPerPage);
+        $pager->setCurrentPage(
+            $request->query->get('page', 1)
```
Comment (emodric) [inline lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Why not transfer the current page with the method parameter to remove dependency on the `Request` object?

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+use eZ\Publish\API\Repository\Values\Content\LocationQuery;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\ContentSearchAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\ContentSearchHitAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\LocationSearchAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\LocationSearchHitAdapter;
+use Pagerfanta\Adapter\AdapterInterface;
+use Pagerfanta\Pagerfanta;
+use Symfony\Component\HttpFoundation\Request;
+
+trait PagerfantaFindServiceSearchAdaptersTrait
```
Comment (emodric) [inline lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: `PagerfantaTrait` ? :)

Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: In future maybe we would like to have trait that is internally using FilterService

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 2 above)
Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: But why ? On all projects we are getting `page` parameter from `Request` object, with this it would be automatically done for you.

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 1 above)
Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Done.

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 3 above)
Comment (emodric) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Then just `PagerfantaFindTrait`?

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * @return Pagerfanta
+     */
+    protected function getPagerfantaPager(AdapterInterface $adapter, Request $request, $maxPerPage)
+    {
+        $pager = new Pagerfanta($adapter);
+        $pager->setNormalizeOutOfRangePages(true);
+        $pager->setMaxPerPage($maxPerPage);
+        $pager->setCurrentPage(
+            $request->query->get('page', 1)
```
Comment (emodric) [inline lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Is it a good idea to have the `Request` here? With this, I can't practically use the trait outside of controller code. It would be better to transfer the current page in method argument.

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 2 above)
Comment (emodric) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Because this way, I can't just simply use the trait to create paged output outside of controller.

What if the query parameter is not named `page`?

What if I want to have two paged queries on the same page, one displaying a query in first page, one displaying a different query on second page?

It's not more difficult at all transfering `$request->query->get('page')` to the method call instead of just `$request`, code wise.

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 3 above)
Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Like it, +1, @pspanja wdyt ?

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 2 above)
Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: I that case you should avoid using this shorthand methods. The idea is to cover 80-90% of use cases with this.

Comment (emodric) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: I disagree. The point of the trait is to be reusable in most usecases and using it outside of a controller where no request is readily available is a valid and often used usecase.

Moreover, using a page query parameter is only an implementation detail coming from Pagerfanta Bundle and not the Pagerfanta lib itself.

Comment (MarioBlazek) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: I've updated code. wdyt @emodric ?

Comment (emodric) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: Looks good :+1:

File: lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php (hunk 3 above)
Comment (pspanja) [reply lib/Core/Traits/PagerfantaFindServiceSearchAdaptersTrait.php]: 👍

File: lib/Core/Traits/PagerfantaFindTrait.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+use Pagerfanta\Pagerfanta;
+
+trait PagerfantaFindTrait
+{
+    /**
+     * Returns Pagerfanta pager that starts from first page
+     * configured with
+     * ContentSearchAdapter and FindService
+     *
+     * @param Query $query
```
Comment (pspanja) [inline lib/Core/Traits/PagerfantaFindTrait.php]: Please use FQCN in docblock hints.

**Comment by pspanja** [comment]
I suggest reducing `create*Pager()` to only two methods, `createSearchPager()` and `createSearchHitPager()`. Inside them we can test given `$query` to decide what adapter to use.

File: lib/Core/Traits/PagerfantaFindTrait.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+     * Shorthand method for creating Pagerfanta pager
+     * with preconfigured Adapter
+     *
+     * @param AdapterInterface $adapter
+     * @param int $currentPage
+     * @param int $maxPerPage
+     *
+     * @return Pagerfanta
+     */
+    protected function getPagerfantaPager(AdapterInterface $adapter, $currentPage, $maxPerPage)
```
Comment (pspanja) [inline lib/Core/Traits/PagerfantaFindTrait.php]: Given we don't have `Pagerfanta` in the method names above, maybe we can drop it here as well.
Also, this method can start with `create` like the ones above, or we can use `new` instead, like is done in Repository API. What do you think?

Comment (emodric) [reply lib/Core/Traits/PagerfantaFindTrait.php]: `getPager` sounds okay. I would keep it at `get*` or `new*` to keep it distinct from `create*` methods.

**Comment by MarioBlazek** [comment]
Please review @pspanja @emodric

File: tests/lib/Integration/Traits/PagerfantaFindTraitTest.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\ContentSearchAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\ContentSearchHitAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\LocationSearchHitAdapter;
+use Netgen\EzPlatformSiteApi\Core\Site\Pagination\Pagerfanta\LocationSearchAdapter;
+use Netgen\EzPlatformSiteApi\Core\Traits\PagerfantaFindTrait;
+use Netgen\EzPlatformSiteApi\Tests\Integration\BaseTest;
+
+class PagerfantaFindTraitTest extends BaseTest
+{
+    use PagerfantaFindTrait;
```
Comment (emodric) [inline tests/lib/Integration/Traits/PagerfantaFindTraitTest.php]: This feels wrong. A test class that uses the trait under test instead of instantiating a (stub) object that uses the trait?

**Comment by MarioBlazek** [comment]
@emodric wdyt ?

**Comment by MarioBlazek** [comment]
Trait methods are using `$this->getSite()->getFindService()` calls.

**Comment by emodric** [comment]
Shouldn't then `getSite` method as well as the `$site` property be defined inside the trait?

**Comment by emodric** [comment]
Or ideally, only the property, without the method at all.

File: UPGRADE.md [hunk 8]
```diff
@@ -1,6 +1,17 @@
 eZ Platform Site API upgrade instructions
 =========================================
 
+Upgrading from 2.2 to 2.3
+-------------------------
+
+Controllers that extend from `Netgen\Bundle\EzPlatformSiteApiBundle\Controller\Controller` and are registered inside dependency injection container must set `parent` config key within service definition, example:
```
Comment (pspanja) [inline UPGRADE.md]: Why is this needed?

File: bundle/Controller/Controller.php [hunk 9]
```diff
@@ -3,19 +3,13 @@
...
 use Netgen\EzPlatformSiteApi\Core\Traits\SearchResultExtractorTrait;
 
 abstract class Controller extends BaseController
 {
     use SearchResultExtractorTrait;
-
-    /**
-     * @return \Netgen\EzPlatformSiteApi\API\Site
-     */
-    public function getSite()
```
Comment (pspanja) [inline bundle/Controller/Controller.php:15]: We can't remove that as this class depends on it in `getRootLocation()`, and `PagerfantaFindTrait` should not be required for it to work.

**Review by pspanja — COMMENTED** [review]
What about using an abstract trait method to implement access to `Site`? I'm aware it's bad from DX side to require additional implementation to use the trait, but not sure how else to solve that properly.

**Comment by pspanja** [comment]
> I suggest reducing create*Pager() to only two methods, createSearchPager() and createSearchHitPager(). Inside them we can test given $query to decide what adapter to use.

What about this one?

File: bundle/Controller/Controller.php (hunk 9 above)
Comment (emodric) [reply bundle/Controller/Controller.php:15]: As I understood @MarioBlazek, one of the points was to use `PagerfantaFindTrait` inside this controller too, in which case this method is not needed any more since it's available inside the trait itself.

File: UPGRADE.md (hunk 8 above)
Comment (emodric) [reply UPGRADE.md]: This would be the solution to your question about abstract trait method. A minor BC break, yes, but it would not require an additional implementation to use the trait, it would already be implemented with the trait and called when instatiating the controller. Hence requirement to use the correct parent in service definition.

Comment (pspanja) [reply UPGRADE.md]: It should be explicitly worded, as far as I understand this is really about requiring `Site` injection through `setSite()` method.

File: bundle/Controller/Controller.php (hunk 9 above)
Comment (pspanja) [reply bundle/Controller/Controller.php:15]: That's kind of obvious, but what if we introduce another trait instead:

```php
SiteAwareTrait
{
    public setSite($site);
    protected getSite();
}
```

Then we can use it to implement abstract trait method `PagerfantaFindTrait::getSite()`.

What do you think?

Comment (MarioBlazek) [reply bundle/Controller/Controller.php:15]: +1 for `SiteAwareTrait`, that was my first idea to go this way.

Comment (emodric) [reply bundle/Controller/Controller.php:15]: `SiteAwareTrait` is fine, but as I understood, you were against it @pspanja ? :)

**Comment by MarioBlazek** [comment]
@pspanja we actually agreed to have explicit methods.

File: bundle/Controller/Controller.php (hunk 9 above)
Comment (pspanja) [reply bundle/Controller/Controller.php:15]: I was? Where?

Comment (MarioBlazek) [reply bundle/Controller/Controller.php:15]: We talked in person :)

Comment (pspanja) [reply bundle/Controller/Controller.php:15]: Oh well :D

**Comment by MarioBlazek** [comment]
@emodric @pspanja can we then summarize what needs to be done ?

**Comment by emodric** [comment]
Tha way it is now, is fine by me, with explicit upgrade instructions about specifying the parent for controller services.

**Comment by MarioBlazek** [comment]
I've updated PR, @pspanja please check it out.

File: UPGRADE.md [hunk 10]
```diff
@@ -1,6 +1,17 @@
 eZ Platform Site API upgrade instructions
 =========================================
 
+Upgrading from 2.2 to 2.3
+-------------------------
```
Comment (emodric) [inline UPGRADE.md]: Needs to be 2.4 now :)

Comment (MarioBlazek) [reply UPGRADE.md]: That's an eagle eye @emodric, updated :)

**Review by emodric — APPROVED** [review]

File: UPGRADE.md [hunk 11]
```diff
@@ -1,6 +1,17 @@
 eZ Platform Site API upgrade instructions
 =========================================
 
+Upgrading from 2.3 to 2.4
+-------------------------
+
+Controllers that extend from `Netgen\Bundle\EzPlatformSiteApiBundle\Controller\Controller` and are registered inside dependency injection container must set `parent` config key within service definition, example:
```
Comment (pspanja) [inline UPGRADE.md]: This is because of `setContainer` and `setSite` calls, right? In that case it's not really a must, I would mention required calls explicitly, and then suggest that it can be done by setting parent configuration which already does it.

Otherwise this looks good :)

**Comment by MarioBlazek** [comment]
@pspanja final updates done.

**Review by pspanja — APPROVED** [review]

**Comment by pspanja** [comment]
🎉 Thanks @MarioBlazek!
